sdks/python/apache_beam/transforms/core.py

```diff
@@ -1300,12 +1300,13 @@ def to_runner_api_parameter(self, context):
           common_urns.requirements.REQUIRES_STATEFUL_PROCESSING.urn)
     from apache_beam.runners.common import DoFnSignature
     sig = DoFnSignature(self.fn)
-    is_splittable = sig.is_splittable_dofn()
```

Comment:
Not sure if checking get_restriction_coder() return type instead of is_splittable_dofn() is future proof.

Also, I don't understand the change, from a mypy correctness perspective.

Reply:
> Not sure if checking get_restriction_coder() return type instead of is_splittable_dofn() is future proof.

`get_restriction_coder()` calls `is_splittable_dofn()` and returns `None` if it's not splittable.  So I interpreted a `None` result from this method to mean "is not splittable". 

```python
  def get_restriction_coder(self):
    # type: () -> Optional[TupleCoder]

    """Get coder for a restriction when processing an SDF. """
    if self.is_splittable_dofn():
      return TupleCoder([
          (self.get_restriction_provider().restriction_coder()),
          (self.get_watermark_estimator_provider().estimator_state_coder())
      ])
    else:
      return None
```

> I don't understand the change, from a mypy correctness perspective.

Here's the problem:

```python
    if is_splittable:
      restriction_coder = sig.get_restriction_coder()  #  returns Optional[TupleCoder]
      restriction_coder_id = context.coders.get_id(restriction_coder)  # does not accept Optional!
    else:	    else:
      restriction_coder_id = None
```

With my changes, we naturally drop the optionality before passing the value to `context.coders.get_id()`.  We also avoid a redundant call to `is_splittable_dofn()`, FWIW.

I see two options:

1) keep my changes and update the documentation of `get_restriction_coder()` to clarify that `None` result indicates "is not splittable"
2) revert my changes and add `assert restriction_coder is None` before the call to `context.coders.get_id()`

Reply:
I don't know why, but `ParDoPayload` below accepts both splittable and restriction_coder_id keywords (theoretically splittable might be True while restriction_coder_id is None), so I think it's safer to do this:
```py
is_splittable = sig.is_splittable_dofn()
restriction_coder = sig.get_restriction_coder()
if restriction_coder:
  restriction_coder_id = context.coders.get_id(
      restriction_coder)  # type: typing.Optional[str]
else:
  restriction_coder_id = None
```

Reply:
It is an error to say is_splittable_dofn is True without returning a restriction coder as well and vice versa.

Reply:
> It is an error to say is_splittable_dofn is True without returning a restriction coder as well and vice versa.

This seems to validate my earlier assessment that a None result from this `get_restriction_coder` means "is not splittable", and therefore that my proposed change is valid.